File: 0962-maximum-width-ramp/0962-maximum-width-ramp.cpp
class Solution {
public:
    int maxWidthRamp(vector<int>& nums) {
        stack<int> st;
        for(int i=0;i<nums.size();i++){
            if(st.empty() || nums[i] <= nums[st.top()]){
                st.push(i);
            }
        }
        int mx = 0;
        for(int j=nums.size()-1;j>=0;j--){
            while(!st.empty() && nums[st.top()] <= nums[j]){
                mx = max(mx,j-st.top());
                st.pop();
            }
        }
        return mx;
    }
};

/*

create a decreasing stack

6,0,8,2,1,5

0
6




*/

File: 0962-maximum-width-ramp/0962-maximum-width-ramp.java
class Solution {
    public int maxWidthRamp(int[] nums) {
        Stack<Integer> st = new Stack<>();
        
        for(int i=0;i<nums.length;i++){
            if(st.isEmpty() || nums[i] <= nums[st.peek()]){
                st.push(i);
            }
        }
        int mx = 0;
        for(int j=nums.length-1;j>=0;j--){
            while(!st.isEmpty() && nums[st.peek()] <= nums[j]){
                mx = Math.max(mx,j-st.peek());
                st.pop();
            }
        }
        return mx;
    }
}